src/dispensary.js

```diff
@@ -0,0 +1,218 @@
+import fs from 'fs';
+
+import async from 'async';
+import request from 'request';
+
+import { DEFAULT_HAHES_FILE, DEFAULT_LIBRARY_FILE } from 'const';
+import hasher from 'hasher';
+import log from 'logger';
+import { getVersions } from 'versions';
+
+
+function fileFormat(url, {file=null, version=null}={}) {
+  if (!file || !version) {
+    throw new Error('ArgumentError: File and version are required.');
+  }
+
+  return url.replace('$VERSION', version)
+            .replace('$FILENAME', file)
+            .replace('$VERSION', version)
+            .replace('$FILENAME', file);
+}
+
+export default class Dispensary {
+
+  constructor(config={}, _libraries=null, _hashes=DEFAULT_HAHES_FILE) {
+    this._libraries = _libraries;
+    this.libraryFile = DEFAULT_LIBRARY_FILE;
+    this.hashesFile = _hashes;
+
+    if (config._ && config._[0]) {
+      this.libraryFile = config._[0];
+    }
+
+    this.maxHTTPRequests = parseInt(config.max);
+  }
+
+  run() {
+    return this.getLibraries()
+      .then((libraries) => {
+        return this.getVersions(libraries);
+      })
+      .then((libraries) => {
+        return this.getFiles(libraries);
+      })
+      .then((libraries) => {
+        return this.getHashes(libraries);
+      })
+      .then((libraries) => {
+        return this.outputHashes(libraries);
+      })
+      .catch((err) => {
+        console.error('ERROR', err);
+      });
+  }
+
+  getLibraries() {
+    if (this._libraries !== null) {
+      return Promise.resolve(this._libraries);
+    }
+
+    try {
+      var libraryJSON = fs.readFileSync(this.libraryFile);
+      this._libraries = JSON.parse(libraryJSON);
+      return Promise.resolve(this._libraries);
+    } catch (err) {
+      if (err.toString().match(/^SyntaxError/)) {
+        return Promise.reject(new Error(
+          `JSONError: ${this.libraryFile} is not valid JSON.`));
+      }
+
+      return Promise.reject(new Error(
+        `${this.libraryFile} does not exist or is not a file.`));
+    }
+  }
+
+  _getAllFilesFromLibrary(library) {
+    var files = [];
+
+    for (let version of library.versions) {
+      if (library.filename) {
+        files.push({
+          file: library.filename,
+          fileOut: library.filenameOutput || library.filename,
+          library: library,
+          version: version,
+        });
+      }
+
+      if (library.filenameMinified) {
+        files.push({
+          file: library.filenameMinified,
+          fileOut: library.filenameMinifiedOutput || library.filenameMinified,
+          library: library,
+          version: version,
+        });
+      }
+    }
+
+    return files;
+  }
+
+  getFiles(libraries) {
+    return new Promise((resolve) => {
+      var files = [];
+
+      var queue = async.queue(this._getFile, this.maxHTTPRequests);
+      queue.drain = function() {
+        log.debug('All downloads completed.');
+        resolve(libraries);
+      };
+
+      for (let i in libraries) {
```

Comment:
Libraries is an array of all of the libraries we're going to generate hashes from.

So libraries is actually an array and `i` is the index. I could do `library of libraries`, but if libraries is for some reason empty/not defined it errors... that shouldn't happen now so I'll see about that, but the `i` here really is a zero-index :)